Extensibility

otrok7 opened the pull request:
I still need to add extensibility to the meeting template and other editors, but I thought I'd create this pull request already, because this feature probably need some discussion.

One of the nice things in BMLT is that it is extensible, the admin can add whatever fields he needs into the meeting definition.  The point of this feature is to add the corresponding flexibility to Bread.

When I look through the code, I see special columns have already been added, eg, Bus_Lines, area_i...in Germany, we also have some special fields.  One alternative would be to add them to the master but this leads to featuritis and a big mess.

What I'm suggesting here is that we allow site admins to add functionality without messing with the core.  A certain amount could be done in configuration...eg, we could let the user enter additional columns in the backend UI, but I would also like to have calculated columns.  For instance, I don't want to use the code for FC3 or LANG formats...I want to print them in the meeting description so that even newcomers (who are the main readers of meeting lists, every else reads them online) can understand.  The LANGs I want by the meeting name, unless the meeting name already has the language (eg, English Meeting).  To handle this sort of logic, I need to be able to add code.

To see an example of how an extension would be programmed, see https://github.com/otrok7/bread-example-extensions

radius314 commented on bmlt-meeting-list.php:
```diff
@@ -22,9 +22,57 @@ class Bread {
 		var $formats_spanish = '';
 		var $formats_all = '';
-		
+		var $meeting_fields = array (
```
Might it make sense to query the root server for available fields rather than hardcode it?

For example: https://example.org/main_server/client_interface/json/?switcher=GetFieldKeys

otrok7 replied:
I thought this was a change you made in 1.10, that the fields retrieved should be limited to the ones really used, for performance reasons.   But if performance is OK getting all meeting data, that's a good idea.

radius314 replied:
Yes, this was.  I wasn't sure as to how you how were pulling "other fields".  The performance is definitely important for larger schedules.  You can leave this then.

radius314 commented on bmlt-meeting-list.php:
```diff
@@ -22,9 +22,57 @@ class Bread {
+			'day_abbr',
+			'area_name',
+		);	
+		var $synonyms = array (
```
might this be better known as `$legacy_synonyms`, since we are keeping this for backward compatibility?

radius314 commented on bmlt-meeting-list.php:
```diff
@@ -434,28 +482,27 @@ function get($url, $cookies = null) {
             return wp_remote_get($url, $args);
 		}
-
-		function get_all_meetings() {
```
is this deliberately commented out?

otrok7 replied:
seems to be dead code,  but I wasn't sure.

radius314 replied:
this is being referenced in partials/_bmlt_server_setup.php, bombing out with an error.

radius314 commented on bmlt-meeting-list.php:
```diff
@@ -1694,7 +1729,23 @@ function write_detailed_formats($formats, $page) {
 		}
-
+		private function parse_field($text)
+        {
```
tabbing issue.

otrok7 commented:
Added the ability to edit the shortcode menu in the tiny_mce editors.  Rather a brute force method, but I tried some others, where the extension provided some PHP structure,and then this is injected into the provided tiny_mce plugin, and this didn't seem to buy any ease of use for the user, and added a lot of code complexity.

radius314 commented:
I'm not sure I understand your last comment.

otrok7 commented:
The other commits in this branch where about extensions in the generation of meeting lists.  This commit is about adding the user-defined shortcodes into the UI, into the editor where the user defines the meeting templates, the front page, the custom section.  And since it's front-end, it needs some javascript, rather than PHP.  And that makes defining the extension awkward.  I could have defined some kind of PHP structure, and generated Javascript out of that, but it seemed to me, it was easier for the user (ie, the admin writing the extension plugin) just to tell him, modify the Javascript yourself.  Look and my sample extension, maybe that helps to understand.

otrok7 commented:
Chris, I asked you to review because this seemed like a large enough change, that I wanted to make sure everyone was on board.

radius314 commented:
I'm not understanding the cause or reason to allow someone to add custom shortcodes. I am not comfortable with introducing such a change as it opens up too many possible issues.

radius314 commented:
Can you explain the reason for allowing someone to add custom shortcodes?  What is the justification of such a feature?

radius314 commented:
FYI, Chris although a major contributor to the BMLT root server (it's a creator), has not delved into the world of "bread".  I don't think it makes sense then to have him review since he's not familiar with the code base.

ChrisMarshallNY commented:
I have always been leery of plugins and extensions that add executable functionality via a regular WP author (as opposed to an admin). Even admins can be questionable, as most blog owners go in as full admins, and couldn't fight their way out of a wet paper bag, technically.

Remember that Danny works in the security industry, so he knows what he's talking about.

This could be one of those "careful what you pray for" things.

Many security breaches come about, because programmers, used to having full authority, write utilities that go in ten levels above what they need.

otrok7 commented:
BMLT is extensible, admins can define new fields of meeting data.  But these fields are not accessible from Bread.  You can see the result in Bread's code...apparently, someone define a column "Bus_Lines" and then convinced you (or whoever was responsible then) to add it in to the core.  This is wasteful for areas that don't have "Bus_lines" defined, but even worse, it means that anyone who defines a new column (and there are a few new columns here in Germany) has to go a convince you to add the column, or you have to tell him what column to use instead (since it's not documented).

Then apparently someone needed the first initial of the area (area_i).  This seems like something very few people need, but there it is, in the core functionality of Bread.  This isn't a new column of meeting data, it's something computed from the meeting data.

This pull request allows users, in particular, the German region, to add their own new fields and computed values.  The alternative would be to add the individual columns I'm using in the German BMLT into Bread, but to me, they seem not so useful for other regions...this way seemed better.

If something seems generally useful, we can always add it in.  If users are happy with Bread without extensions, they don't even know they are there.  But if someone calls and says 'I have a field that that describes where to park, or whatever, or where the location info's english translation is..., then we can give them an immediate solution....

Does that motivate you?

radius314 commented:
@otrok7 every field that is available in the root server should be retrievable through this endpoint https://example.org/main_server/client_interface/json/?switcher=GetFieldKeys.  Pull that and populate a new submenu called "Custom", which would solve every use case.

radius314 commented:
I agree with the premise of not adding a field for every case, it was design decision I was not a part of when it was made and there for legacy reasons.  Once this custom feature is implemented, we can probably pull back a lot of those custom fields.

otrok7 commented:
It would be technically possible to remove some of the fields being retrieved and shortcodes being computed.  Whether that's a good practice depends on your users.  But I don't want to be the one making it even worse.

radius314 commented:
That's not what I was referring to.  Can you explain why dynamically pulling the fields that are non-standard from the root server are not sufficient?

otrok7 commented:
shit, just realized I missed the SCRUM.  Sorry.

radius314 commented:
Scrum is next sunday.

radius314 commented:
9 am NY time

otrok7 commented:
You could pull the fields, but then you run into performance problems, like we talked about before.  Our server is regional, the meeting lists are from the areas...pulling the fields would return the superset of fields for all areas.  

And don't forget we also have some computed fields, like area_i in Bread now, or the language notation exactly when the language is not in the name of the meeting.  Or getting Helpline details from the service body, which is a current issue.

radius314 commented:
You do not need to pull all the fields, you can pull fields that are "in-use", that match the list of custom fields.  This is not an all or nothing solution.  Computed fields should continue to operate as they have been.  Helpline details are not part of meeting list information, this is a standard field that is part of the schema of service bodies and should not be included in this conversation.

otrok7 commented:
Let me make sure I understand what you're suggesting: we get the list of fields from getFieldKeys, then offer them in the meeting template.  When we generate, we look at the template first, and try to identify all keys being used, and then generate our query based on that.  In the UI we put all the fields we don't know about in a section called custom, and forget about placing them correctly in the menu.

Is that right?

What about computed fields that we in germany are using?   And the extensibility system allows shortcodes to be added to the custom_section, or front_page, or last_page, too.  I need this, e.g., because I need to switch from a 4-column portrait layout to a 3 column landscape layout (see https://na-berlin.de/?current-meeting-list=2), and I can well imagine that someone will want to display information about the service-areas or helplines there.

radius314 commented:
The line of thinking is correct, however, I do not understand what you mean by "forget about placing them correctly in the menu".  What do you mean by correctly?  

Can you define which are the special computed fields you are using?

otrok7 commented:
By correctly, I meant here that we can't know if they should be in the "location" or "info" part of the menu.  But this is really not so important.

Computed fields:  
1. We don't like using symbols for format, we write them out.  And in different places (languages by the name, as described below), FC1s underneath, FC3s more prominant, etc.

2. Meetings want their FC1 formats on the meeting list.  If the meeting has a complicated format then officially the meeting has a rotating format, but in the meeting list we dont print that, but rather a new field, "Format-Explanation".  Not printing the "rotating format" saves space, which is an issue when then meeting list should fit on one page.  What sort of complicated formats do I have to deal with:  Here's and example: week 1: speaker, week 2; It Works (step of the month) 3: It works (Trad of the month) 4; Other Lit --- believe it or not Germans love thinking this stuff up and seeing it written down!

3. We also have logic that controls the names.  If the name is the default name (NA Meeting) and it's in another language, or special interest I use that to generate a name.  If it does have a name and is in another language, then I tack on the language (or languages) 

The whole point is that I don't think this kind of logic should be in Bread...I don't even think it should be in Berlin...but pleasing the area, and later the region, and being able to make the meeting lists to their specification was absolutely essential to getting Bread, and BMLT accepted and used.

otrok7 commented:
I need to sleep now... if you answer and I don't respond, it's because I've gone to bed.

ChrisMarshallNY commented:
Sleep well.

I must say, no matter how this moves forward, I'm really glad to have you helping us out Ron.

Thanks so much for stepping up. You are helping to save many lives; not just in Germany.

radius314 commented:
1) I think language is implemented incorrectly in bread, this was never refactored.  IMO, language should be a query string parameter, rather than something that is tied to a given configuration.

What are FC1 and FC3?  I'm not familiar with this terminology?

ChrisMarshallNY commented:
Those are the original format “category” codes I put in the format record.

These are “broad categories” of formats, such as ones that affect who can attend, discussion types, accessibility, etc.

They never ended up being used, but there is a slot in the format record for them.

radius314 commented:
I see, is there a way to pull those out using a semantic endpoint?

I tried, but it doesn't seem to output.

https://example.org/main_server/client_interface/json/?switcher=GetFormats

ChrisMarshallNY commented:
Maybe not. I think I stopped using them when 2.0 came out, and that was when I created the semantic API.

otrok7 commented:
Format type codes are back, see root server commit 173.
    Am Montag, 17. Juni 2019, 03:11:43 MESZ hat Chris Marshall <notifications@github.com> Folgendes geschrieben:  
 
 
Maybe not. I think I stopped using them when 2.0 came out, and that was when I created the semantic API.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

ChrisMarshallNY commented:
I just put these links on Slack:

wip branch (not yet integrated): http://tkddevel.com/bmltwork/bmlt-wip/main_server/client_interface/json/?switcher=GetFormats

Current release branch: https://bmlt.newyorkna.org/main_server/client_interface/json/?switcher=GetFormats

otrok7 commented:
Do you think that fc1_keys, fc1_names, fc1_descr, fc2_keys, fc2_names, fc2_descr etc, etc should be added to Bread?

radius314 commented:
My idea was that someone could make a custom computed field by creating a tinymce editor and point to any available set of data endpoints, save it and then reference it in any of the other places( front page, meetings etc)

radius314 commented:
This would be secure and avoid unwanted support maintenance.

otrok7 commented:
What are your security concerns?  Bread is a client, and users can write their own clients, they can't do anything in Bread extensions that they can't do themselves, right?

How would you express something like "if the FC1 format of the meeting is 'Rotating', and there is the 'format_comments' field is non-blank, then print the format_comments field, otherwise, print the fc1 format of the meeting?"  I think if we could express something like that in a shortcode, we'd both be happy.  The trick is not to wind up inventing our own programming language.  There's a sweet spot somewhere there.

otrok7 commented:
And of course, the whole point of extensions is to avoid support and maintenance.  Because otherwise you get people from other regions, who want Bread to strange things, and you have to build them into your core.  People like me, I mean.  But also, whoever requested "detailed_format_codes_spanish" etc.

radius314 commented:
I am going to leave this request open, and pause this, until I have more time to devote to this.

otrok7 commented:
I wanted to propose the same...it's not like I think that a solution requiring the users to program PHP is optimal....

But there is actually a lot that we agree on: that there is a need for extensibility, and that the refactorings in this pull request are steps in this direction... I think that the only uncertainty is whether the extensions come through an "apply_filters" or through some other mechanism.

With that in mind my suggestion would be to let me comment the apply_filter lines out, and then approve the pull request.   This would let me go ahead and implement the last change I talked about in the beginning, generalizing the ASM mechanism;  You should be able to apply a template to ASM meetings, too (currently the template is hardcoded!), and any format should be selectable as the format that appears in the list (currently is is hardcoded as format "ASM").

Is that OK for you?

otrok7 commented:
This article makes a nice case for this type of extensibility
https://www.smashingmagazine.com/2018/03/making-wordpress-plugin-extensible/

radius314 commented:
I have merged this but something doesn't seem right.  If you look at this root server https://bmlt.sezf.org/main_server/client_interface/csv/?switcher=GetFieldKeys, the field `train_lines` exists, however the short code doesn't seem to be dynamically getting pulled in.

radius314 commented:
Looks like the javascript is not being set correctly.

radius314 commented:
The right way to approach this is to have an endpoint that emotes Javascript and ultimately acts like the `front_page_button` tinyMCE plugin.

otrok7 commented:
Hi Danny,
the way it's currently coded, nothing is automatically pulled in.  Fields are added (or removed) has to be added in through the extension plugin, but even that is currently commented out: I thought you wanted to have more time to think about an alternate mechanism.
In the case of the fields being retrieved, it's pretty easy to imagine a multiple select (using Chosen, of course!) to identify the interesting fields. 
Or we can just pull in all fields, with get_field_keys.  But that has performance costs (have you actually measured this? it's pretty surprising that this is a noticeable difference given how deadly slow mpdf is).  Or do you have something in mind as a way to identify interesting (and non-private) fields?
I think both are OK ideas...then will limit the PHP based extensibility to things actually requiring coding.
Ron
    Am Donnerstag, 27. Juni 2019, 03:57:24 MESZ hat Danny Gershman <notifications@github.com> Folgendes geschrieben:  
 
 
I have merged this but something doesn't seem right. If you look at this root server https://bmlt.sezf.org/main_server/client_interface/csv/?switcher=GetFieldKeys, the field train_lines exists, however the short code doesn't seem to be dynamically getting pulled in.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

otrok7 commented:
That was my first thought, that we generate the javascript based on field names (including computed fields).  But if you want to give the user the flexibility to arrange the fields how he wants, including sections for meeting location, area, etc., then you can't do so much better than the format tinyMCE already defines. 
But yeah, this is something to work on.
    Am Donnerstag, 27. Juni 2019, 04:22:41 MESZ hat Danny Gershman <notifications@github.com> Folgendes geschrieben:  
 
 
The right way to approach this is to have an endpoint that emotes Javascript and ultimately acts like the front_page_button tinyMCE plugin.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

otrok7 commented:
Worst bug report ever ;-).  What seems to be the problem?
    Am Donnerstag, 27. Juni 2019, 04:14:07 MESZ hat Danny Gershman <notifications@github.com> Folgendes geschrieben:  
 
 
Looks like the javascript is not being set correctly.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.
